jormungandr/src/utils/task.rs

```diff
@@ -132,31 +143,48 @@ impl Services {
             handle,
         };
 
-        let finish_notifier = self.finish_listener.notifier();
-        self.runtime.spawn(async move {
+        let handle = self.runtime.spawn(async move {
             let res = f(future_service_info).await;
-            let outcome = if res.is_ok() {
-                "successfully"
+            if let Err(err) = &res {
+                error!(logger, "service finished with error"; "reason" => err.to_string());
             } else {
-                "with error"
-            };
-            info!(logger, "service finished {}", outcome);
-
-            // send the finish notifier if the service finished with an error.
-            // this will allow to finish the node with an error code instead
-            // of an success error code
-            let _ = finish_notifier.sender.send(res.is_ok());
-            // Holds finish notifier, so it's dropped when whole future finishes or is dropped
-            std::mem::drop(finish_notifier);
+                info!(logger, "service finished successfully");
+            }
+            res.map_err(Into::into)
         });
+        self.finish_listener.push(handle);
 
         let task = Service::new(name, now);
         self.services.push(task);
     }
 
     /// select on all the started services. this function will block until first services returns
-    pub fn wait_any_finished(&self) -> Result<bool, RecvError> {
-        self.finish_listener.wait_any_finished()
+    pub fn wait_any_finished(mut self) -> Result<(), ServiceError> {
+        let finish_listener = self.finish_listener;
+        let result = self
+            .runtime
+            .block_on(async move { finish_listener.into_future().await.0 });
+        match result {
+            // No services were started or some service exited successfully
+            None | Some(Ok(Ok(()))) => Ok(()),
+            // Error produced by a service
+            Some(Ok(Err(service_error))) => Err(ServiceError::Service(service_error)),
+            // A service panicked or was cancelled by the environment
+            Some(Err(join_error)) => {
+                if join_error.is_cancelled() {
+                    Err(ServiceError::Cancelled)
+                } else if join_error.is_panic() {
+                    let desc = join_error
+                        .into_panic()
+                        .downcast_ref::<String>()
+                        .expect("a panic must be convertible to String")
```

Comment:
This is not necessarily true: `panic!` can box anything as the value.
I think it's better to make the `ServiceError::Panic` variant type `Option<String>`.